et_replay/comm/backend/pytorch_dist_backend.py

```diff
@@ -1070,9 +1073,29 @@ def initialize_backend(
     def initialize_groups(self, backend="gloo"):
         groups = {}
         world_size = self.get_world_size()
+        global_rank = self.get_global_rank()
+
+        # sync pgs across ranks to fix hang with multiple comm groups
+        # because new_group() functions requires that all processes in the main group enter,
+        # even if they are not going to be members of the group.
+        # Assumption: pg_name is unique and consistent for all ranks
+        sync_store = dist.PrefixStore("pg_sync_r", self.tcp_store)
+        sync_store.set(str(global_rank), json.dumps(self.commsParams.groupRanks))
+        torch.distributed.barrier()
+        group_ranks_sync = self.commsParams.groupRanks.copy()
+        for i in range(self.get_world_size()):
+            if i == global_rank:
+                continue
+            json_data = sync_store.get(str(i))
+
+            # convert pg_id in json_data to int
+            pg_id2group_ranks = {}
+            for pg_id, group_ranks in json.loads(json_data).items():
+                pg_id2group_ranks[int(pg_id)] = group_ranks
+            group_ranks_sync.update(pg_id2group_ranks)
 
         # create additional groups
-        for pg_id, group_ranks in self.commsParams.groupRanks.items():
+        for pg_id, group_ranks in dict(sorted(group_ranks_sync.items())).items():
```

Comment:
Observation by @GSSBMW: "`sorted` is needed to keep all ranks use the same order to create all process groups"





[6:17](https://nvidia.slack.com/archives/D07EM2GU9D0/p1722939460141499)
to keep all ranks use the same order to create all process groups